File: test/test_reduce.cpp
#include <stdio.h>
#include "TRTCContext.h"
#include "DVVector.h"
#include "reduce.h"

int main()
{
	TRTCContext ctx;
	{
		int harr[6] = { 1, 0, 2, 2, 1, 3 };
		DVVector darr(ctx, "int32_t", 6, harr);

		ViewBuf ret;
		TRTC_Reduce(ctx, darr, ret);
		printf("%d\n", *(int*)ret.data());

		TRTC_Reduce(ctx, darr, DVInt32(1), ret);
		printf("%d\n", *(int*)ret.data());

		TRTC_Reduce(ctx, darr, DVInt32(-1), Functor("Maximum"), ret);
		printf("%d\n", *(int*)ret.data());
	}

	{
		int h_keys_in[7] = { 1, 3, 3, 3, 2, 2, 1 };
		DVVector d_keys_in(ctx, "int32_t", 7, h_keys_in);
		int h_values_in[7] = { 9, 8, 7, 6, 5, 4, 3 };
		DVVector d_values_in(ctx, "int32_t", 7, h_values_in);

		int h_keys_out[7];
		DVVector d_keys_out(ctx, "int32_t", 7);
		int h_values_out[7];
		DVVector d_values_out(ctx, "int32_t", 7);

		unsigned count = TRTC_Reduce_By_Key(ctx, d_keys_in, d_values_in, d_keys_out, d_values_out);
		d_keys_out.to_host(h_keys_out, 0, count);
		d_values_out.to_host(h_values_out, 0, count);
		for (unsigned i = 0; i < count; i++)
			printf("%d ", h_keys_out[i]);
		puts("");
		for (unsigned i = 0; i < count; i++)
			printf("%d ", h_values_out[i]);
		puts("");

	}

	{
		int h_keys_in[7] = { 1, 3, 3, 3, 2, 2, 1 };
		DVVector d_keys_in(ctx, "int32_t", 7, h_keys_in);
		int h_values_in[7] = { 9, 8, 7, 6, 5, 4, 3 };
		DVVector d_values_in(ctx, "int32_t", 7, h_values_in);

		int h_keys_out[7];
		DVVector d_keys_out(ctx, "int32_t", 7);
		int h_values_out[7];
		DVVector d_values_out(ctx, "int32_t", 7);

		unsigned count = TRTC_Reduce_By_Key(ctx, d_keys_in, d_values_in, d_keys_out, d_values_out, Functor("EqualTo"));
		d_keys_out.to_host(h_keys_out, 0, count);
		d_values_out.to_host(h_values_out, 0, count);
		for (unsigned i = 0; i < count; i++)
			printf("%d ", h_keys_out[i]);
		puts("");
		for (unsigned i = 0; i < count; i++)
			printf("%d ", h_values_out[i]);
		puts("");

	}

	{
		int h_keys_in[7] = { 1, 3, 3, 3, 2, 2, 1 };
		DVVector d_keys_in(ctx, "int32_t", 7, h_keys_in);
		int h_values_in[7] = { 9, 8, 7, 6, 5, 4, 3 };
		DVVector d_values_in(ctx, "int32_t", 7, h_values_in);

		int h_keys_out[7];
		DVVector d_keys_out(ctx, "int32_t", 7);
		int h_values_out[7];
		DVVector d_values_out(ctx, "int32_t", 7);

		unsigned count = TRTC_Reduce_By_Key(ctx, d_keys_in, d_values_in, d_keys_out, d_values_out, Functor("EqualTo"), Functor("Plus"));
		d_keys_out.to_host(h_keys_out, 0, count);
		d_values_out.to_host(h_values_out, 0, count);
		for (unsigned i = 0; i < count; i++)
			printf("%d ", h_keys_out[i]);
		puts("");
		for (unsigned i = 0; i < count; i++)
			printf("%d ", h_values_out[i]);
		puts("");

	}

	return 0;
}

File: CSharp/test/test_reduce/test_reduce.cs
using System;
using ThrustRTCSharp;

namespace test_reduce
{
    class test_reduce
    {
        static void Main(string[] args)
        {
            {
                DVVector darr = new DVVector(new int[] { 1, 0, 2, 2, 1, 3 });
                Console.WriteLine(TRTC.Reduce(darr));
                Console.WriteLine(TRTC.Reduce(darr, new DVInt32(1)));
                Console.WriteLine(TRTC.Reduce(darr, new DVInt32(-1), new Functor("Maximum")));
            }

            {
                DVVector d_keys_in = new DVVector(new int[] { 1, 3, 3, 3, 2, 2, 1 });
                DVVector d_values_in = new DVVector(new int[] { 9, 8, 7, 6, 5, 4, 3 });
                DVVector d_keys_out = new DVVector("int32_t", 7);
                DVVector d_values_out = new DVVector("int32_t", 7);
                long count = TRTC.Reduce_By_Key(d_keys_in, d_values_in, d_keys_out, d_values_out);
                print_array((int[])d_keys_out.to_host(0, count));
                print_array((int[])d_values_out.to_host(0, count));
            }

            {
                DVVector d_keys_in = new DVVector(new int[] { 1, 3, 3, 3, 2, 2, 1 });
                DVVector d_values_in = new DVVector(new int[] { 9, 8, 7, 6, 5, 4, 3 });
                DVVector d_keys_out = new DVVector("int32_t", 7);
                DVVector d_values_out = new DVVector("int32_t", 7);
                long count = TRTC.Reduce_By_Key(d_keys_in, d_values_in, d_keys_out, d_values_out, new Functor("EqualTo"));
                print_array((int[])d_keys_out.to_host(0, count));
                print_array((int[])d_values_out.to_host(0, count));
            }

            {
                DVVector d_keys_in = new DVVector(new int[] { 1, 3, 3, 3, 2, 2, 1 });
                DVVector d_values_in = new DVVector(new int[] { 9, 8, 7, 6, 5, 4, 3 });
                DVVector d_keys_out = new DVVector("int32_t", 7);
                DVVector d_values_out = new DVVector("int32_t", 7);
                long count = TRTC.Reduce_By_Key(d_keys_in, d_values_in, d_keys_out, d_values_out, new Functor("EqualTo"), new Functor("Plus"));
                print_array((int[])d_keys_out.to_host(0, count));
                print_array((int[])d_values_out.to_host(0, count));
            }
        }

        static void print_array<T>(T[] arr)
        {
            foreach (var item in arr)
            {
                Console.Write(item.ToString());
                Console.Write(" ");
            }
            Console.WriteLine("");
        }
    }
}
